nibabel/streamlines/array_sequence.py

```diff
@@ -0,0 +1,267 @@
+
+        # Clear unused memory.
+        if self._data.ndim != 0:
+            self._data.resize((offset,) + self.common_shape)
+
+    @property
+    def is_array_sequence(self):
+        return True
+
+    @property
+    def common_shape(self):
+        """ Matching shape of the elements in this array sequence. """
+        if self._data.ndim == 0:
+            return ()
+
+        return self._data.shape[1:]
+
+    def append(self, element):
+        """ Appends :obj:`element` to this array sequence.
+
+        Parameters
+        ----------
+        element : ndarray
+            Element to append. The shape must match already inserted elements
+            shape except for the first dimension.
+
+        Notes
+        -----
+        If you need to add multiple elements you should consider
+        `ArraySequence.extend`.
+        """
+        if self._data.ndim == 0:
+            self._data = np.asarray(element).copy()
+            self._offsets = np.array([0])
+            self._lengths = np.array([len(element)])
+            return
+
+        if element.shape[1:] != self.common_shape:
+            msg = "All dimensions, except the first one, must match exactly"
+            raise ValueError(msg)
+
+        self._offsets = np.r_[self._offsets, len(self._data)]
+        self._lengths = np.r_[self._lengths, len(element)]
+        self._data = np.append(self._data, element, axis=0)
+
+    def extend(self, elements):
+        """ Appends all `elements` to this array sequence.
+
+        Parameters
+        ----------
+        elements : list of ndarrays or :class:`ArraySequence` object
+            If list of ndarrays, each ndarray will be concatenated along the
+            first dimension then appended to the data of this ArraySequence.
+            If :class:`ArraySequence` object, its data are simply appended to
+            the data of this ArraySequence.
+
+        Notes
+        -----
+            The shape of the elements to be added must match the one of the
+            data of this :class:`ArraySequence` except for the first dimension.
+        """
+        if len(elements) == 0:
+            return
+
+        if self._data.ndim == 0:
+            elem = np.asarray(elements[0])
+            self._data = np.zeros((0, elem.shape[1]), dtype=elem.dtype)
+
+        next_offset = self._data.shape[0]
+
+        if is_array_sequence(elements):
+            self._data.resize((self._data.shape[0] + sum(elements._lengths),
+                               self._data.shape[1]))
+
+            offsets = []
+            for offset, length in zip(elements._offsets, elements._lengths):
+                offsets.append(next_offset)
+                chunk = elements._data[offset:offset + length]
+                self._data[next_offset:next_offset + length] = chunk
+                next_offset += length
+
+            self._lengths = np.r_[self._lengths, elements._lengths]
+            self._offsets = np.r_[self._offsets, offsets]
+
```

Comment:
I refactored the entire method. It takes less lines and now supports `elements` to be a generator, yeah! It calls the `ArraySequence`'s constructor if `elements` is not an `ArraySequence` object before recalling `self.extend` with that new instance.